server/test/list_objects.go

```diff
@@ -0,0 +1,185 @@
+	"github.com/openfga/openfga/pkg/logger"
+	"github.com/openfga/openfga/pkg/telemetry"
+	"github.com/openfga/openfga/pkg/testutils"
+	"github.com/openfga/openfga/server/commands"
+	"github.com/openfga/openfga/storage"
+	teststorage "github.com/openfga/openfga/storage/test"
+	"github.com/stretchr/testify/require"
+	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+const (
+	defaultListObjectsDeadline   = 5 * time.Second
+	defaultListObjectsMaxResults = 3
+)
+
+type listObjectsTestCase struct {
+	_name          string
+	request        *openfgapb.ListObjectsRequest
+	expectedError  error
+	expectedResult []string //all the results. the server may return less
+}
+
+var tkAnnaRepo1 = &openfgapb.TupleKey{
+	Object:   "repo:1",
+	Relation: "admin",
+	User:     "anna",
+}
+var tkAnnaRepo2 = &openfgapb.TupleKey{
+	Object:   "repo:2",
+	Relation: "admin",
+	User:     "anna",
+}
+var tkAnnaRepo3 = &openfgapb.TupleKey{
+	Object:   "repo:3",
+	Relation: "admin",
+	User:     "anna",
+}
+var tkAnnaRepo4 = &openfgapb.TupleKey{
+	Object:   "repo:4",
+	Relation: "admin",
+	User:     "anna",
+}
+var tkBobRepo4 = &openfgapb.TupleKey{
+	Object:   "repo:2",
+	Relation: "admin",
+	User:     "bob",
+}
+
+func newListObjectsRequest(store, objectType, relation, user, modelID string) *openfgapb.ListObjectsRequest {
+	return &openfgapb.ListObjectsRequest{
+		StoreId:              store,
+		Type:                 objectType,
+		Relation:             relation,
+		User:                 user,
+		AuthorizationModelId: modelID,
+	}
+}
+
+func TestListObjects(t *testing.T, dbTester teststorage.DatastoreTester[storage.OpenFGADatastore]) {
+	store := testutils.CreateRandomString(10)
+	tracer := telemetry.NewNoopTracer()
+	ctx, backend, modelID, err := setupTestListObjects(store, dbTester)
+	require.NoError(t, err)
+
+	t.Run("list objects", func(t *testing.T) {
+		testCases := []listObjectsTestCase{
+			{
+				_name:          "does not return duplicates and respects maximum length allowed",
+				request:        newListObjectsRequest(store, "repo", "admin", "anna", modelID),
+				expectedResult: []string{"1", "2", "3", "4"},
+				expectedError:  nil,
+			},
+			{
+				_name:          "performs correct checks",
+				request:        newListObjectsRequest(store, "repo", "admin", "bob", modelID),
+				expectedResult: []string{"2"},
+				expectedError:  nil,
+			},
+		}
+
+		listObjectsQuery := commands.NewListObjectsQuery(backend, tracer, logger.NewNoopLogger(), telemetry.NewNoopMeter(), defaultListObjectsDeadline, defaultListObjectsMaxResults, defaultResolveNodeLimit)
+		runListObjectsTests(t, ctx, testCases, listObjectsQuery)
+	})
+}
+
+func runListObjectsTests(t *testing.T, ctx context.Context, testCases []listObjectsTestCase, listObjectsQuery *commands.ListObjectsQuery) {
+	var res *openfgapb.ListObjectsResponse
+	var err error
+	for _, test := range testCases {
+		res, err = listObjectsQuery.Execute(ctx, test.request)
```

Comment:
This file only tests the non-streaming version... still need to find out if there is a way of testing the streaming version.

Reply:
Consume the stream in its entirety and then make your test assertions. That's how we should test the streaming version.

Reply:
I'm not sure how to instantiate the stream :/ @jon-whit do you mind sending a follow-up PR for these tests? Or if you can point me to an example?

Reply:
See:
https://grpc.io/docs/languages/go/basics/#client-side-streaming-rpc
https://grpc.io/docs/languages/go/basics/#client-side-streaming-rpc